(MFTF)MSI mftf credit memo tests

magterskine opened the pull request:
### Description (*)
Credit memo tests for MSI
#1949 : CreditMemoCreatedForWholeOrderWithSimpleProductOnDefaultStockAfterFullInvoiceAndShipmentInAdminTest
#1973 : AdminCreditMemoCreatedWithPartialRefundForOrderWithSimpleProductOnDefaultStockAfterFullInvoiceInAdminTest
~#1974 :  CreditMemoCreatedWithFullRefundWithSimpleProductOnDefaultStockAfterFullInvoiceAndPartialShipmentTest~
#1976 : CreditMemoCreatedWithPartialRefundWithSimpleProductOnDefaultStockAfterFullInvoiceAndPartialShipment


### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

magterskine committed:
MSI test MSI-1949

magterskine committed:
MSI-1949: Removing unnecessary waits

magterskine committed:
magento-engcom/msi#2.3-develop-MFTF-CreditMemo

MSI-1949 - Credit Memo test

magterskine committed:
magento-engcom/msi#2.3-develop-MFTF-CreditMemo

added MSI-1973: Credit memo test with partial refund

magterskine committed:
Merge branch '2.3-develop' into 2.3-develop-MFTF-CreditMemo

magterskine committed:
magento-engcom/msi#2.3-develop-MFTF-CreditMemo

New test
MSI-1974: CreditMemoCreatedWithFullRefundWithSimpleProductOnDefaultStockAfterFullInvoiceAndPartialShipmentTest

magterskine committed:
magento-engcom/msi#1949

MSI-1949, MSI-1973, MSI-1974
Removed unneccessary comments from tests.

okolesnyk commented:
Build link: https://m2build-ur.devops.magento.com/job/All-User-Requested-Tests/11779/

okolesnyk commented on app/code/Magento/InventoryAdminUi/Test/Mftf/Test/AdminCreditMemoCreatedForWholeOrderWithSimpleProductOnDefaultStockAfterFullInvoiceAndShipmentInAdminTest.xml:
```diff
@@ -0,0 +1,180 @@
+        </annotations>
+
+        <before>
+            <magentoCLI stepKey="enableStockManagement" command="config:set cataloginventory/item_options/manage_stock 1"/>
```
This option is enabled by default in Magento, so there is no need to add this command to before.

okolesnyk commented on app/code/Magento/InventoryAdminUi/Test/Mftf/Test/AdminCreditMemoCreatedForWholeOrderWithSimpleProductOnDefaultStockAfterFullInvoiceAndShipmentInAdminTest.xml:
```diff
@@ -0,0 +1,180 @@
+        <click selector="{{StorefrontMinicartSection.itemQuantityUpdate($$simpleProduct1.name$$)}}" stepKey="updateQtyInMiniCart"/>
+        <click selector="{{StorefrontMinicartSection.goToCheckout}}" stepKey="goToCheckout"/>
+        <waitForPageLoad stepKey="waitForPaymentSelectionPageLoad"/>
+        <click selector=".continue" stepKey="clickOnNext1"/>
```
Please put this element selector to appropriate section or find existing one.

magterskine committed:
magento-engcom/msi#1976

MSI-1976:CreditMemoCreatedWithPartialRefundWithSimpleProductOnDefaultStockAfterFullInvoiceAndPartialShipment

okolesnyk commented on app/code/Magento/InventoryAdminUi/Test/Mftf/Test/AdminCreditMemoCreatedForWholeOrderWithSimpleProductOnDefaultStockAfterFullInvoiceAndShipmentInAdminTest.xml:
```diff
@@ -0,0 +1,180 @@
+                <requiredEntity createDataKey="createSource1"/>
+            </createData>
+            <createData entity="SimpleSubCategory" stepKey="simpleCategory1"/>
+            <createData entity="SimpleMsiProduct" stepKey="simpleProduct1"/>
```
Please pass `SimpleSubCategory` as `<requiredEntity createDataKey="SimpleSubCategory"/>` and Simple product will be persisted with assigned category.

okolesnyk commented on app/code/Magento/InventoryAdminUi/Test/Mftf/Test/AdminCreditMemoCreatedForWholeOrderWithSimpleProductOnDefaultStockAfterFullInvoiceAndShipmentInAdminTest.xml:
```diff
@@ -0,0 +1,180 @@
+        <before>
+            <magentoCLI stepKey="enableStockManagement" command="config:set cataloginventory/item_options/manage_stock 1"/>
+            <createData entity="MsiCustomer1" stepKey="createCustomer1"/>
+            <createData entity="BasicMsiStockWithMainWebsite1" stepKey="createStock1"/>
```
Based on HipTest you should create Stock without specifying Sales Channels.
If you will create Stock without Sales Channel `Main Website` you will not need to perform actions under this test block: `<!--Admin - Assign Default Stock to Main Website-->`

okolesnyk commented on app/code/Magento/InventoryAdminUi/Test/Mftf/Test/AdminCreditMemoCreatedForWholeOrderWithSimpleProductOnDefaultStockAfterFullInvoiceAndShipmentInAdminTest.xml:
```diff
@@ -0,0 +1,180 @@
+        <selectOption selector="{{AdminEditStockSalesChannelsSection.websites}}" userInput="Main Website" stepKey="selectDefaultWebsiteAsSalesChannelForDefaultStock"/>
+        <click selector="{{AdminGridMainControls.saveAndContinue}}" stepKey="saveDefaultStock"/>
+
+        <comment userInput="Assign category to created simple product." stepKey="assignCategoryToProductComment"/>
```
Assign category to product steps must be removed after you will pass `SimpleSubCategory` as `<requiredEntity createDataKey="SimpleSubCategory"/>` to createData action of `SimpleMsiProduct `

okolesnyk commented on app/code/Magento/InventoryAdminUi/Test/Mftf/Test/AdminCreditMemoCreatedForWholeOrderWithSimpleProductOnDefaultStockAfterFullInvoiceAndShipmentInAdminTest.xml:
```diff
@@ -0,0 +1,180 @@
+        <click selector="{{AdminProductGridSection.productGridXRowYColumnButton('1', '2')}}" stepKey="openProductEditPage"/>
+        <scrollToTopOfPage stepKey="scrollToTopOfPage1"/>
+        <searchAndMultiSelectOption selector="{{AdminProductFormSection.categoriesDropdown}}" parameterArray="[$$simpleCategory1.name$$]" requiredAction="true" stepKey="searchAndSelectCategory"/>
+        <fillField selector="{{AdminProductSourcesGrid.rowQty('0')}}" userInput="1000" stepKey="fillSourceQtyField"/>
```
Based on Hiptest Default Source qty is 100. Please make data in Hiptest and Automated test consistent.

okolesnyk commented on app/code/Magento/InventoryAdminUi/Test/Mftf/Test/AdminCreditMemoCreatedForWholeOrderWithSimpleProductOnDefaultStockAfterFullInvoiceAndShipmentInAdminTest.xml:
```diff
@@ -0,0 +1,180 @@
+        <comment userInput="Purchase 10 simple product" stepKey="purchaseSimpleProductComment"/>
+        <amOnPage url="{{StorefrontCategoryPage.url($$simpleCategory1.name$$)}}" stepKey="navigateToCategoryPage"/>
+        <moveMouseOver selector="{{StorefrontCategoryProductSection.ProductInfoByName($$simpleProduct1.name$$)}}" stepKey="moveMouseOverProduct" />
+        <click selector="{{StorefrontCategoryProductSection.ProductAddToCartByName($$simpleProduct1.name$$)}}" stepKey="clickAddToCart" />
```
Based on HipTest Product should be added to shopping cart from Product Page with Qty of 5 products.

okolesnyk commented on app/code/Magento/InventoryAdminUi/Test/Mftf/Test/AdminCreditMemoCreatedForWholeOrderWithSimpleProductOnDefaultStockAfterFullInvoiceAndShipmentInAdminTest.xml:
```diff
@@ -0,0 +1,180 @@
+        <amOnPage url="{{AdminOrdersPage.url}}" stepKey="onOrdersPageToCreateShipment"/>
+        <waitForLoadingMaskToDisappear stepKey="waitForOrdersPageLoadingMask"/>
+        <fillField selector="{{AdminOrdersGridSection.search}}" userInput="{$grabOrderNumber}"
+                   stepKey="searchOrderNum"/>
```
Please write this action in one line and verify all other actions are using similar style

okolesnyk commented on app/code/Magento/InventoryAdminUi/Test/Mftf/Test/AdminCreditMemoCreatedForWholeOrderWithSimpleProductOnDefaultStockAfterFullInvoiceAndShipmentInAdminTest.xml:
```diff
@@ -0,0 +1,180 @@
+        <click selector="{{AdminOrderDetailsMainActionsSection.ship}}" stepKey="clickShip"/>
+        <waitForLoadingMaskToDisappear stepKey="waitForShipLoadingMask"/>
+        <click selector="{{AdminShipmentMainActionsSection.submitShipment}}" stepKey="submitShipment"/>
+        <see selector="{{AdminOrderDetailsMessagesSection.successMessage}}" userInput="The shipment has been created." stepKey="seeShipmentCreateSuccess"/>
```
Please also verify this steps:
```
Verify Shipment is present in "Shipments" tab of Order
Verify that your Order got status "Processing"
```

okolesnyk commented on app/code/Magento/InventoryAdminUi/Test/Mftf/Test/AdminCreditMemoCreatedForWholeOrderWithSimpleProductOnDefaultStockAfterFullInvoiceAndShipmentInAdminTest.xml:
```diff
@@ -0,0 +1,180 @@
+        <waitForPageLoad time="60" stepKey="waitForPageLoad21"/>
+        <scrollToTopOfPage stepKey="scrollToTopMessage"/>
+        <waitForPageLoad stepKey="waitForPageLoad22"/>
+        <see selector="{{AdminOrderDetailsMessagesSection.successMessage}}" userInput="The invoice has been created." stepKey="checkSuccessMessage"/>
```
Please also verify
```
Verify Invoice is present in "Invoices" tab of Order
Verify that your Order got status "Processing"
```

okolesnyk reviewed (changes requested):
1. Please review all `stepKey` attributes
 - do not use numbers for stepKeys
 - some stepKeys are not describing real action. Example: `        <see selector="{{CheckoutPaymentSection.billingAddress}}" userInput="{{US_Address_TX.street[0]}}" stepKey="chooseBillingAddress"/>
`
2. Based on comments to first test make similar changes to second and third test.
3. Keep HipTest and Automated test consistent in all ways (Data, Flow, etc.)
4. Mark HipTests as partially automated or discuss with MSI team how you should treat test which contains verification of data in Database cause MFTF is not capable of doing such things.

okolesnyk commented on app/code/Magento/InventoryAdminUi/Test/Mftf/Test/AdminCreditMemoCreatedForWholeOrderWithSimpleProductOnDefaultStockAfterFullInvoiceAndShipmentInAdminTest.xml:
```diff
@@ -0,0 +1,180 @@
+        <waitForPageLoad stepKey="waitForPageLoad22"/>
+        <see selector="{{AdminOrderDetailsMessagesSection.successMessage}}" userInput="The invoice has been created." stepKey="checkSuccessMessage"/>
+
+
```
Please get rid of extra empty lines

okolesnyk commented on app/code/Magento/InventoryAdminUi/Test/Mftf/Test/AdminCreditMemoCreatedForWholeOrderWithSimpleProductOnDefaultStockAfterFullInvoiceAndShipmentInAdminTest.xml:
```diff
@@ -0,0 +1,180 @@
+</tests>
+
+
+
```
Please get rid of extra empty lines

magterskine committed:
MSI Pull Request 1975 - fixing issues based on code review

magterskine committed:
Merge branch '2.3-develop' into 2.3-develop-MFTF-CreditMemo

magterskine commented:
Build : https://m2build-ur.devops.magento.com/job/All-User-Requested-Tests/12993/

magterskine committed:
Modifying groups and suite file to remove CreditMemo group used for local verification

magterskine committed:
Merge branch '2.3-develop-MFTF-CreditMemo' of https://github.com/magento-engcom/msi into 2.3-develop-MFTF-CreditMemo

magterskine committed:
Due to UI changes - updates to test and Section on selectors for ProductGrid/QuantityBySource

magterskine committed:
CreditMemoCreatedWithFullRefundWithSimpleProductOnDefaultStockAfterFullInvoiceAndPartialShipmentTest - fixed incorrect source and stock values in final assert

magterskine committed:
Removing   CreditMemoCreatedWithFullRefundWithSimpleProductOnDefaultStockAfterFullInvoiceAndPartialShipmentTest.xml

magterskine committed:
Remove whitespace

magterskine committed:
Updating Selector to remove strong tag

magterskine committed:
Merge branch '2.3-develop' into 2.3-develop-MFTF-CreditMemo

magterskine committed:
Fixing suite file which had local testing suites

magterskine committed:
Merge branch '2.3-develop-MFTF-CreditMemo' of https://github.com/magento-engcom/msi into 2.3-develop-MFTF-CreditMemo

magterskine committed:
Merge branch '2.3-develop' into 2.3-develop-MFTF-CreditMemo

magterskine commented:
UR Build: https://m2build-ur.devops.magento.com/job/All-User-Requested-Tests/16907/
All added tests passed in CE/EE/B2B

ghost commented:
Hi @magterskine, thank you for your contribution! </br>Please, complete [Contribution Survey](https://opensource.magento.com/survey/contribution/magento-engcom+msi+1975), it will take less than a minute. </br>Your feedback will help us to improve contribution process.

okolesnyk commented on app/code/Magento/InventoryAdminUi/Test/Mftf/Test/AdminCreditMemoCreatedForWholeOrderWithSimpleProductOnDefaultStockAfterFullInvoiceAndShipmentInAdminTest.xml:
```diff
@@ -0,0 +1,149 @@
+            <createData entity="SourceStockLinked1" stepKey="linkSourceStock">
+                <requiredEntity createDataKey="createStock"/>
+                <requiredEntity createDataKey="createSource"/>
+            </createData>
```
<!-- Assign Sales Channel to Default Stock -->
            <amOnPage url="{{AdminManageStockPage.url}}" stepKey="amOnTheStockGridPage"/>
            <waitForPageLoad time="30" stepKey="waitForStockGridPageLoad"/>
            <actionGroup ref="searchAdminDataGridByKeyword" stepKey="searchDefaultStockByKeyword">
                <argument name="keyword" value="_defaultStock.name"/>
            </actionGroup>
            <click selector="{{AdminGridRow.editByValue(_defaultStock.name)}}"
                   stepKey="clickOnEditForDefaultStock"/>
            <waitForPageLoad time="30" stepKey="waitForStockEditPageLoad"/>
            <selectOption selector="{{AdminEditStockSalesChannelsSection.websites}}" userInput="Main Website"
                          stepKey="selectWebsiteAsSalesChannel"/>
            <click selector="{{AdminGridMainControls.saveAndContinue}}" stepKey="saveDefaultStock"/>
